ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -280,6 +293,115 @@ public synchronized Container getContainerById(int accountId, int containerId) t
+   * @return the {@link Dataset} which contains the metadata.
+   * @throws SQLException
+   */
+  public synchronized Dataset addDatasetVersions(int accountId, int containerId, String accountName,
+      String containerName, String datasetName, long version, long expirationTimeMs) throws SQLException {
+    Dataset dataset;
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      dataAccessor.getDatabaseConnection(true);
+      PreparedStatement insertDatasetVersionStatement =
+          dataAccessor.getPreparedStatement(insertDatasetVersionSql, true);
+      dataset = executeAddDatasetVersionStatement(insertDatasetVersionStatement, accountId, containerId, accountName,
+          containerName, datasetName, version, expirationTimeMs);
+      dataAccessor.onSuccess(Write, System.currentTimeMillis() - startTimeMs);
+      return dataset;
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Write);
+      throw e;
+    }
+  }
+
+  /**
+   * Get a version of {@link Dataset} based on supplied properties.
+   * @param accountId the id for the parent account.
+   * @param containerId the id of the container.
+   * @param datasetName the name of the dataset.
+   * @param version the version of the dataset.
+   * @return the {@link DatasetVersionRecord}
+   * @throws SQLException
+   */
+  public synchronized DatasetVersionRecord getDatasetVersions(short accountId, short containerId, String datasetName,
+      long version) throws SQLException {
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      dataAccessor.getDatabaseConnection(false);
+      PreparedStatement getDatasetVersionStatement =
+          dataAccessor.getPreparedStatement(getDatasetVersionByNameSql, false);
+      DatasetVersionRecord result =
+          executeGetDatasetVersionStatement(getDatasetVersionStatement, accountId, containerId, datasetName, version);
+      dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
+      return result;
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Read);
+      throw e;
+    }
+  }
+
+  /**
+   * Execute insert dataset version statement to add a dataset version.
+   * @param statement the mysql statement to add dataset version.
+   * @param accountId the id for the parent account.
+   * @param containerId the id of the container.
+   * @param datasetName the name of the dataset.
+   * @param version the version of the dataset.
+   * @param expirationTimeMs the expiration time in milliseconds.
+   * @throws SQLException
+   */
+  private Dataset executeAddDatasetVersionStatement(PreparedStatement statement, int accountId, int containerId,
+      String accountName, String containerName, String datasetName, long version, long expirationTimeMs)
+      throws SQLException {
+    Dataset dataset = getDatasetMetadata(accountId, containerId, accountName, containerName, datasetName, expirationTimeMs);
```

Comment:
nit: Since `getDatasetMetadata()` itself is getting data from MySql, would it better to call in caller function `addDatasetVersions` so that `executeAddDatasetVersionStatement` is only executing prepared statement?